src/initialization/MOM_state_initialization.F90

```diff
@@ -1821,73 +1826,82 @@ subroutine initialize_sponges_file(G, GV, US, use_temperature, tv, param_file, C
 
-  if ( GV%nkml>0 .and. .not. new_sponges) then
-    ! This call to set_up_sponge_ML_density registers the target values of the
-    ! mixed layer density, which is used in determining which layers can be
-    ! inflated without causing static instabilities.
-    do i=is-1,ie ; pres(i) = tv%P_Ref ; enddo
-    EOSdom(:) = EOS_domain(G%HI)
-
-    call MOM_read_data(filename, potemp_var, tmp(:,:,:), G%Domain)
-    call MOM_read_data(filename, salin_var, tmp2(:,:,:), G%Domain)
 
-    do j=js,je
-      call calculate_density(tmp(:,j,1), tmp2(:,j,1), pres, tmp_2d(:,j), tv%eqn_of_state, EOSdom)
-    enddo
+    ! The remaining calls to set_up_sponge_field can be in any order.
+    if ( use_temperature) then
+      call MOM_read_data(filename, potemp_var, tmp(:,:,:), G%Domain)
+      call set_up_sponge_field(tmp, tv%T, G, nz, Layer_CSp)
+      call MOM_read_data(filename, salin_var, tmp(:,:,:), G%Domain)
+      call set_up_sponge_field(tmp, tv%S, G, nz, Layer_CSp)
+    endif
 
-    call set_up_sponge_ML_density(tmp_2d, G, CSp)
   endif
 
-  ! The remaining calls to set_up_sponge_field can be in any order.
-  if ( use_temperature .and. .not. new_sponges) then
-    call MOM_read_data(filename, potemp_var, tmp(:,:,:), G%Domain)
-    call set_up_sponge_field(tmp, tv%T, G, nz, CSp)
-    call MOM_read_data(filename, salin_var, tmp(:,:,:), G%Domain)
-    call set_up_sponge_field(tmp, tv%S, G, nz, CSp)
-  elseif (use_temperature) then
-    call set_up_ALE_sponge_field(filename, potemp_var, Time, G, GV, US, tv%T, ALE_CSp)
-    call set_up_ALE_sponge_field(filename, salin_var, Time, G, GV, US, tv%S, ALE_CSp)
+
+  if  (use_ALE) then
+    if (.not. time_space_interp_sponge) then ! ALE mode
+      call field_size(filename,eta_var,siz,no_domain=.true.)
+      if (siz(1) /= G%ieg-G%isg+1 .or. siz(2) /= G%jeg-G%jsg+1) &
+        call MOM_error(FATAL,"initialize_sponge_file: Array size mismatch for sponge data.")
+      nz_data = siz(3)-1
+      allocate(eta(isd:ied,jsd:jed,nz_data+1))
+      allocate(h(isd:ied,jsd:jed,nz_data))
+
+      call MOM_read_data(filename, eta_var, eta(:,:,:), G%Domain, scale=US%m_to_Z)
+
+      do j=js,je ; do i=is,ie
+        eta(i,j,nz+1) = -G%bathyT(i,j)
+      enddo ; enddo
+
+      do k=nz,1,-1 ; do j=js,je ; do i=is,ie
+        if (eta(i,j,K) < (eta(i,j,K+1) + GV%Angstrom_Z)) &
+          eta(i,j,K) = eta(i,j,K+1) + GV%Angstrom_Z
+      enddo ; enddo ; enddo
+      do k=1,nz; do j=js,je ; do i=is,ie
+        h(i,j,k) = GV%Z_to_H*(eta(i,j,k)-eta(i,j,k+1))
+      enddo ; enddo ; enddo
+      call initialize_ALE_sponge(Idamp, G, param_file, ALE_CSp, h, nz_data)
+      deallocate(eta)
+      deallocate(h)
+    else
+      ! Initialize sponges without supplying sponge grid
+      call initialize_ALE_sponge(Idamp, G, param_file, ALE_CSp)
+    endif
+
+
+    ! The remaining calls to set_up_sponge_field can be in any order.
+    if ( use_temperature) then
+      call set_up_ALE_sponge_field(filename, potemp_var, Time, G, GV, US, tv%T, ALE_CSp)
```

Comment:
set_up_ALE_sponge_field is an overloaded interface to two different routines. I would have thought that the two different interfaces should have been used for the two settings of time_space_interp_sponge, but both options are using the variant that should be used with time_space_interp_sponge=.true.

  Given that this code was not working before, in a way that our testing was not detecting, I think that we need to be particularly careful in implementing these corrections. As a confidence-building measure, I would prefer that this be implemented in the simpler way, using the static target value interface when time_space_interp_sponge=.false. Once this has been demonstrated to work (e.g., the MOM6-issue that triggered all this has been closed), we could then try using the same interface for both variants and verifying that they give identical answers in test cases that exercise that option.

Reply:
@Hallberg-NOAA . Updated as suggested.

Reply:
I am now satisfied that with these latest modifications this PR is very likely to be exactly what we need to address and close Issue #1218.